src/epoch_credits/mod.rs

```diff
@@ -1,35 +1,176 @@
+            let (staking_apys, validator_rewards) =
+                self.process_rewards(rewards, epochs_in_year)?;
+
+            for s in staking_apys {
+                self.staking_apys
+                    .get_metric_with_label_values(&[&s.voter])
+                    .map(|c| c.set(s.percent))?;
+            }
+
+            for v in validator_rewards {
+                self.validator_rewards
+                    .get_metric_with_label_values(&[&v.voter])
+                    .map(|c| c.set(v.lamports as f64 / LAMPORTS_PER_SOL as f64))?;
+            }
+            // FIXME: add to seen epochs in the cache.
+            self.last_rewards_epoch = epoch;
+        }
+        Ok(())
+    }
+
+    /// Splits rewards into reward type categories and does post-processing.
+    fn process_rewards(
+        &self,
+        rewards: Rewards,
+        epochs_in_year: f64,
+    ) -> anyhow::Result<(Vec<StakingApy>, Vec<ValidatorReward>)> {
+        let mut staking_seen_voters = BTreeSet::new();
+        let mut staking_apys = Vec::new();
+        let mut validator_rewards = Vec::new();
+
+        for Reward {
+            pubkey,
+            lamports,
+            post_balance,
+            reward_type,
+        } in rewards
+        {
+            match reward_type {
+                Some(RewardType::Staking) => {
+                    let account_info = self
+                        .client
+                        .get_account(&Pubkey::try_from(pubkey.as_ref())?)?;
+                    let stake_state: StakeState = bincode::deserialize(&account_info.data)?;
+                    if let Some(delegation) = stake_state.delegation() {
+                        let voter = format!("{}", delegation.voter_pubkey);
+                        if !staking_seen_voters.contains(&voter) && lamports > 0 {
+                            let lamports = lamports as u64;
+                            let prev_balance = post_balance - lamports;
+                            let epoch_rate = lamports as f64 / prev_balance as f64;
+                            let apy = 100.0
+                                * (f64::powf(1.0 + epoch_rate / epochs_in_year, epochs_in_year)
+                                    - 1.0);
+                            staking_apys.push(StakingApy {
+                                voter: voter.clone(),
+                                percent: apy,
+                            });
+                            staking_seen_voters.insert(voter);
+                        }
+                    }
+                }
+                Some(RewardType::Voting) => validator_rewards.push(ValidatorReward {
+                    voter: pubkey,
+                    lamports: post_balance,
+                }),
+                _ => (), // TODO other reward types
+            }
+        }
+        Ok((staking_apys, validator_rewards))
+    }
+
+    /// Gets the rewards for `epoch` optionally given `epoch_info`. Returns `Ok(None)` if there
```

Comment:
I don't quite understand why the function allows an `EpochInfo` to be passed in, defaulting to the RPC's answer on `None`. Wouldn't this be unnecessary since we do not expect `EpochInfo` to change at all?

Reply:
Good point. This is a refactoring artefact.